### ISSUE TITLE
Generate TypeScript files instead of JS files?

### ISSUE THREAD

**Issue by sbking** [opral/inlang-paraglide-js#238]
I can not enable `allowJs` for my application because other build tooling (Remix + Vite) does not work well when `allowJs` is enabled. Is there any way to make Paraglide generate .ts files instead?

**Comment by samuelstroschein** [comment]
We can likely offer an option `emitWithTsExtension`

Can you test if changing the file endings to `.ts` works seamlessly in your project? Due to the JSDoc annotations, I expect TypeScript to work out of the box. If that's the case, then we don't have to use the typescript compiler which we tried to avoid to keep the complexity low

**Comment by samuelstroschein** [comment]
Can confirm. TypeScript complains with "ts 80004: Please convert JSDoc to TypeScript types". 

Ergh bummer. It seems like we need to run the typescript compiler with emitDeclarations only to parse the JSDoc and emit d.ts files [https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html). 

Good news: We won't need an option because we can emit both `.js` and `.d.ts` files. JS projects shouldn't be affected. 

* Paraglide keeps being written in JSDoc


* JS projects use JSDoc (.js files) 
* TS projects use declaration files (.d.ts) 

@sbking are you open for a PR against the `lix-integration` branch that has paraglide js 2.0? Then we can include this in the paraglide js 2.0 release

**Comment by samuelstroschein** [comment]
@jeiea, can you elaborate on the issue is you are running into? what do you mean with the ts language server is screwed up when modifying source files?

**Comment by samuelstroschein** [comment]
@jeiea can you check if this also happens if you keep the `.js` files and add a `.d.ts` file next to them?

**Comment by jeiea** [comment]
It looks good, but again, I'm not sure. I can't create a 16000 line d.ts file.

**Comment by samuelstroschein** [comment]
@jeiea okay, i assume it will work. I'm adding this issue to v2.0

**Comment by jeiea** [comment]
d.ts files don't seem to support auto import, is that okay?

**Comment by samuelstroschein** [comment]
You mean having auto-complete that leads to an "auto import"? 

```
happ...

auto completes to
 
happy_elephant()

and adds auto import 

import { happy_elephant } from "paraglide-js/messages.js"
```

**Comment by jeiea** [comment]
Yes. I meant that.

**Comment by jeiea** [comment]
I didn't test against the PR, i can't sure dts has problem or not now because I could make mistake when make it.

I don't think virtual modules is completely better solution, because current inlang compilation can be slow (my another issue), and I'm avoiding it by not running that compillation. I can't sure virtual modules allows that.

**Comment by samuelstroschein** [comment]
Makes sense. Keeping both issues then. Users can opt for virtual modules or only emitting declarations files as they seem fit.

**Comment by samuelstroschein** [comment]
@skbing https://github.com/opral/inlang-paraglide-js/issues/273 should solve this issue

**Comment by samuelstroschein** [comment]
Paraglide JS 2.0 will have an `experimentalEmitTsDeclaration` flag. During the development of the flag it became clear that generating TS files directly is the way to go. I opened https://github.com/opral/inlang-paraglide-js/issues/299 as follow up

**Comment by samuelstroschein** [comment]
Argh... the experimental emitTs directly flag leads to too much overhead. 

- Paraglide needs to rewrite paths
- Every function is implemented twice in JSDoc and TS
- Writing adapters requires dual declaration of JSDoc and TS as well

I considered emitting only TS but that would negatively impact the majority of projects: 

- build time get's longer (which can be an issue see #313) because TS files need to be compiled into JS
- virtual modules would need to compile TS to JS as well

Current thinking is: 

1. Keep emitting JSDoc because it works for >90% of use cases, is simpler, plus runs anywhere (it's just JS)
2. Re-introduce the `emitTsDeclarations` flag for rare cases where direct TS usage is needed

Thoughts?

**Comment by minht11** [comment]
I have been using `experimentalEmitTs` and I liked it, but for my cases `emitTsDeclarations` would also be enough.

**Comment by samuelstroschein** [comment]
`emitTsDeclarations` is doable. Slows down the compiler by 2x but if you can't set `allowJs` in your project, it's reasonable.

**Comment by samuelstroschein** [comment]
With node v24 we could switch to a TS output with [erasable syntax](https://nodejs.org/api/typescript.html#type-stripping) instead of emitting JS + JSDoc. 

- no dual compilation target of JS and/or TS, which leads to [too much maintenance effort](https://github.com/opral/inlang-paraglide-js/issues/238#issuecomment-2603277565)
- works in any JS project because node can read the ts files with node v24 and up
- compile times are not increased because the ts compiler is not used

**Comment by minht11** [comment]
Upgrading from the beta, missing declarations support is the main blocker for us.
I imagine JS/TS should still be configurable option because not many projects will be able to upgrade to node v24 at first.

**Comment by gvillo** [comment]
Wondering how I can get rid of `TS7016: Could not find a declaration file for module zala/paraglide/messages.` without using a `@ts-ignore` on every import, node v24 will become LTS in 6 months :/

**Comment by samuelstroschein** [comment]
@gvillo use `allowJs: true` and `checkJs: true` in your tsconfig

**Comment by lmorais-dev** [comment]
> [@gvillo](https://github.com/gvillo) use `allowJs: true` and `checkJs: true` in your tsconfig

Does not work sadly

**Comment by marviobezerra** [comment]
I'm also having a similar issue. `allowJs: true` and `checkJs: true` don't work.

I'd love to see an option to select the desired output: `JS` or `TS`. I have two projects running on Bun that are 100% TS files.

**Comment by samuelstroschein** [comment]
Are you all compiling paraglide into a sub-package and then import it from your main package? That's the only reasonable explanation for the missing declarations that comes to mind. 

It's best if someone can open a new issue with a reproduction

**Comment by marviobezerra** [comment]
Yes, that is my case. I have a PNPM monorepo where one of the packages is shared between front end and backend. The shared package contains zod schemas and paraglide messages.

**Comment by samuelstroschein** [comment]
Run `tsc` after compiling. TypeScript supports JSDoc to TS Declarations out of the box https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html

**Comment by marviobezerra** [comment]
That is an extra step that could be avoided by generating TS instead of JS.

Is there any technical reason to not generate TS files?

**Comment by marviobezerra** [comment]
Maybe I didn’t make it clear, I asked if it’s possible to generate Typescript instead of JavaScript. The comment you linked talks about generating JavaScript + Type definitions. Which are two different things.

You can think about it as a new generator flavour. The user can pick Typescript or JavaScript as output.

**Comment by samuelstroschein** [comment]
I understood your question correctly. I explored this in https://github.com/opral/inlang-paraglide-js/issues/238#issuecomment-2603277565.

The compiler would need to maintain two compilation targets. I tried it. The overhead is too high. Furthermore, most people want #423 anyway. That completely bypasses emitting files to disk, except for declaration files. 

Now that I think about it, compiling messages into a sub-package to be consumed by an app also seems like an anti-pattern. An inlang project has no dependencies. You can have an inlang project anywhere in your monorepo, and compile the project in your consuming app. That would give you build time optimizations like message modules and potentially virtual modules that you would otherwise miss out on. 

_Both app-a and app-b can compile the same inlang project without compiling the project into a sub-package. The need for a "translation" package does not exist._

```diff
.
├── project.inlang
└── packages/
    ├── app-a
    ├── app-b
-   └── translations
```

**Comment by kaelhem** [comment]
Hi,

I'm following the official guide to set up a Vite project with Paraglide (here: https://inlang.com/m/gerre34r/library-inlang-paraglideJs/vite), but I'm running into issues with TypeScript support.

When I try to import messages like this:
```
import { m } from '@/paraglide/messages'
```

I get the following error:
`Cannot find module '@/paraglide/messages' or its corresponding type declarations.`

As a result, there’s no type checking or autocompletion for messages, which makes development quite frustrating.

To make things easier to reproduce, I’ve created a minimal repo that shows the issue:
https://github.com/kaelhem/vite-paraglide-minimal

Any help would be greatly appreciated. Thanks!

**Comment by kaelhem** [comment]
Thanks for your quick response!

I actually tried adding allowJs and checkJs to tsconfig.node.json, but it didn’t change anything.
However, you pointed me in the right direction — the actual issue was that my tsconfig.app.json was missing this part:
```
"baseUrl": ".",
"paths": {
  "@/*": ["./src/*"]
}
```
Once I added that, everything worked as expected :)

Since these files are the default ones generated by Vite (with the react-ts template), it might be helpful to mention this in the documentation.

Thanks again!

**Comment by samuelstroschein** [comment]
Released in v2.7.0
